docs/proposals/optional_application_loging.md

```diff
@@ -0,0 +1,106 @@
+Optional Application Logging
+============================
+
+The EFK aggregated logging stack by default collects application logs with
+fluentd from STDOUT of each container managed by OpenShift. Currently only way
+of opting out is based on a node label and ansible variable
+`openshift_logging_fluentd_nodeselector` as described in the
+[documentation](https://docs.openshift.org/latest/install_config/aggregate_logging.html).
+This proposal discusses some ways of a finer grained level of control over
+which container logs are stored in Elasticsearch.
+
+
+## Fluentd pipeline
+
+The container logs can be obtained from one of these supported logging drivers
+- json-file - logs read from `/var/log/containers/*` mounted in fluentd
+              container
+- journald
+
+Latest development shifted recommended docker driver from journald to json-file,
+therefore, the proposed implementation has focused on usage of json-file as
+well. Rationale being, journald currently doesn't match performance expectations
+and desired throttling control for larger clusters.
+
+Each container on the host has its own json structured log file
+- `/var/log/containers/[pod]_[namespace]_[container]-[docker_id].log`
+
+Every line is treated as a separate log message and processed in the cascade
+of fluentd filters with currently only supported sink Elasticsearch.
+A log line doesn't contain OpenShift metadata, this is added to the message
+in the fluentd `k8_metadata_filter` plugin.
+
+#### Dynamic Control of Optional Logging
+
+By 'dynamic control' we understand the ability of an operator or a user to
+opt-in or opt-out certain OpenShift object from aggregated logging without
+having to redeploy fluentd.
+
+This approach has immediate effect, offers very granular level of control,
+is flexible, and doesn't bring any measurable latency for other containers logs.
```

Comment:
...doesn't bring any measurable latency...    do we know this enough to be able to state it here?

Reply:
I should probably elaborate what I meant by that statement. I wanted to use it in comparison with static control and emphasize the impact on all containers during any change of settings.

the way I imagine static control is:
- set [`exclude_path`](https://docs.fluentd.org/v0.12/articles/in_tail#exclude_path) to something like `/var/log/containers/mypod_test_test-*.log`
- only place, where this takes effect is here
  - https://github.com/fluent/fluentd/blob/v0.12.41/lib/fluent/plugin/in_tail.rb#L167
- we need to redeploy fluentd to acknowledge new settings
- during redeployment, there is a brief period of time when fluentd doesn't collect logs even from  pods not excluded => a change of settings introduces a spike in latency for other containers

to put it in contrast with dynamic control:
- set label `aggregated_logging_opt_out=true` to a container
- the modified fluentd already queries OpenShift API regardless
- it doesn't need to be redeployed => a change of settings doesn't introduce a spike in latency for other containers

I will try to rephrase the original statement to explain this more clearly